### PR TITLE
EPL-1302: [How-to] How to add a canvas field to a Form or Grid

### PR DESCRIPTION
**Pull request by RomanMagnoli**

### CODE REVIEW

**Review by valenvivaldi — CHANGES_REQUESTED** [review]
Quitar referencia a OB

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      "web/org.openbravo.client.application.examples/js/example-canvas-field.js", false));
+.....
+```
+
+###  Defining the canvas in the Tab-Field (ADField)
+
+The next step is to create a new field in the tab, and set its client class
+field:  
+
+![](/assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-1.png)
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:151]:
```suggestion
![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-1.png)
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+It is also possible to add in-line properties in the "Clientclass" field, for example, 'OBEXAPP_SalesOrderActionButton {"title":
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:153]:
```suggestion
It is also possible to add in-line properties in the `"Clientclass"` field, for example, 'OBEXAPP_SalesOrderActionButton {"title":
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+"My Action Button"}'
+
+###  The result
+
+The result is shown in both the grid and the form:
+
+![](/assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-2.png)
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:160]:
```suggestion
![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-2.png)
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+![](/assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-3.png)
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:162]:
```suggestion
![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-3.png)
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  * When displaying a row in the grid 
+  * When editing a row in the grid 
+  * In form view, when editing a record 
+
+The last 2 cases are similar.
+
+In grid-display mode the following applies:
+
+  * In grid mode the following properties are set on the canvas: 
+    * grid: the grid object
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:178]:
```suggestion
    * `grid`: the grid object 
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  * When editing a row in the grid 
+  * In form view, when editing a record 
+
+The last 2 cases are similar.
+
+In grid-display mode the following applies:
+
+  * In grid mode the following properties are set on the canvas: 
+    * grid: the grid object 
+    * rowNum: the row number/record index for which the canvas is used
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:179]:
```suggestion
    * `rowNum`: the row number/record index for which the canvas is used 
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  * In form view, when editing a record 
+
+The last 2 cases are similar.
+
+In grid-display mode the following applies:
+
+  * In grid mode the following properties are set on the canvas: 
+    * grid: the grid object 
+    * rowNum: the row number/record index for which the canvas is used 
+    * record: the record for which the canvas is used
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:180]:
```suggestion
    * `record`: the record for which the canvas is used 
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+The last 2 cases are similar.
+
+In grid-display mode the following applies:
+
+  * In grid mode the following properties are set on the canvas: 
+    * grid: the grid object 
+    * rowNum: the row number/record index for which the canvas is used 
+    * record: the record for which the canvas is used 
+    * colNum: the column in which the canvas is shown
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:181]:
```suggestion
    * `colNum`: the column in which the canvas is shown 
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+The last 2 cases are similar.
+
+In grid-display mode the following applies:
+
+  * In grid mode the following properties are set on the canvas: 
+    * grid: the grid object 
+    * rowNum: the row number/record index for which the canvas is used 
+    * record: the record for which the canvas is used 
+    * colNum: the column in which the canvas is shown 
+    * field: the field where the canvas is used
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:182]:
```suggestion
    * `field`: the field where the canvas is used 
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    * field: the field where the canvas is used 
+  * A canvas can be created and pooled, so over time it is created once and re-used when records are scrolled into view. 
+  When a canvas is used for a record, then the setRecord method on it is called (if the canvas has this method). 
+
+In form-edit or grid-edit mode, the following applies:
+
+  * The canvasItem property will be set pointing to the form item the canvasItem can be used to get to the form itself and to the grid (if editing in the grid): 
+    * `this.canvasItem.form`: the form instance 
+    * `this.canvasItem.form.grid`: in case of grid editing 
+  * If the canvas has a property noTitle with the value true then in form mode no title/label is displayed
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:191]:
```suggestion
  * If the canvas has a property noTitle with the value `true` then in form mode no title/label is displayed 
```

**Review by valeg-etendo — CHANGES_REQUESTED** [review]

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md (hunk 2 above)
Comment (RomanMagnoli) [reply docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md]: Fixed

**Review by valeg-etendo — CHANGES_REQUESTED** [review]

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+  - JavaScript Development
+  - UI Components
+  - JavaScripts Events
+---
+
+#  How to add a canvas field to a Form or Grid
+  
+##  Introduction
+
+A canvas field allows the user to add any visual component to a form or a row in a
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:16]: Delete line break

**Review by valeg-etendo — CHANGES_REQUESTED** [review]

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+  * [Client Side Development and API](../concepts/Client_Side_Development_and_API.md)
+  * [JavaScript Coding Conventions](../concepts/JavaScript_Coding_Conventions) 
+
+It also makes sense to study the following page: [Etendo Architecture](../concepts/Etendo_Architecture).
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-0.png)
+
+##  Example Module
+
+This section is supported by an example module which shows examples of the code
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:38]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+shown and discussed.
+
+The code of the example module can be downloaded from this
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:41]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+public List<ComponentResource> getGlobalComponentResources() {
+  final List<ComponentResource> globalResources = new ArrayList<ComponentResource>();
+  globalResources.add(createStaticResource(
+      "web/org.openbravo.client.application.examples/js/example-canvas-field.js", false));
+.....
+```
+
+###  Defining the canvas in the Tab-Field (ADField)
+
+The next step is to create a new field in the tab, and set its client class
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:148]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+
+The result is shown in both the grid and the form:
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-2.png)
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid-3.png)
+
+##  Javascript events passed to the canvas
+
+When implementing your canvas, you should take into account that the canvas is
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:166]: Delete line break

**Review by valeg-etendo — CHANGES_REQUESTED** [review]

**Review by valenvivaldi — APPROVED** [review]

**Review by isaiasb-etendo — CHANGES_REQUESTED** [review]

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+  - Form Customization
+  - Grid Customization
+  - JavaScript Development
+  - UI Components
+  - JavaScripts Events
+---
+
+#  How to add a canvas field to a Form or Grid
+  
+##  Introduction
```
[the remark this answers is not in the document]
Comment (RomanMagnoli) [reply docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md:19]: Fix. Added above

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+
+  * Create a  [javascript](../concepts/Client_Side_Development_and_API#adding-javascript-to-etendo) file with your javascript class and place it in the correct directory, the convention is to place js files in the following directory in your module: web/[module.java.package]/js 
+  * Register the javascript file (and other static resources such as css files) in Etendo using a [ComponentProvider](../concepts/Etendo_Architecture/#component-provider) 
+
+The example module implements a button and a calculated field in the `example-canvas-field.js` file.
+
+The first example in the file implements a button which shows the identifier of the current record:
+
+    
+```javascript
```
[the remark this answers is not in the document]
Comment (RomanMagnoli) [reply docs/developer-guide/etendo-classic/how-to-guides/How_to_add_a_canvas_field_to_a_Form_or_Grid.md]: Fixed

**Review by isaiasb-etendo — APPROVED** [review]

**Pull request merged**